src/V3Width.cpp

```diff
@@ -2150,14 +2155,40 @@ class WidthVisitor final : public VNVisitor {
         // Make sure dtype is sized
         nodep->dtypep(iterateEditMoveDTypep(nodep, nodep->subDTypep()));
         UASSERT_OBJ(nodep->dtypep(), nodep, "No dtype determined for var");
-        if (const AstUnsizedArrayDType* const unsizedp
-            = VN_CAST(nodep->dtypeSkipRefp(), UnsizedArrayDType)) {
-            if (!(m_ftaskp && m_ftaskp->dpiImport())) {
-                UINFO(9, "Unsized becomes dynamic array " << nodep << endl);
-                AstDynArrayDType* const newp
-                    = new AstDynArrayDType{unsizedp->fileline(), unsizedp->subDTypep()};
-                nodep->dtypep(newp);
-                v3Global.rootp()->typeTablep()->addTypesp(newp);
+        if (!(m_ftaskp && m_ftaskp->dpiImport())) {
+            AstNodeDType *dtp = nodep->dtypep();
+            AstNodeDType *np = nullptr;
+            while (VN_IS(dtp->skipRefp(), UnsizedArrayDType)
+                    || VN_IS(dtp->skipRefp(), UnpackArrayDType)
+                    || VN_IS(dtp->skipRefp(), AssocArrayDType)
+                    || VN_IS(dtp->skipRefp(), WildcardArrayDType)
+                    || VN_IS(dtp->skipRefp(), QueueDType)) {
+                if (const AstUnsizedArrayDType* const unsizedp
+                    = VN_CAST(dtp->skipRefp(), UnsizedArrayDType)) {
+                    UINFO(9, "Unsized becomes dynamic array ");
+                    if (!np) {
+                        // for Var itself
+                        UINFONL(9, nodep << endl);
+                    } else {
+                        // for subDType
+                        UINFONL(9, dtp << endl);
+                    }
+                    AstDynArrayDType* const newp
+                        = new AstDynArrayDType{unsizedp->fileline(), unsizedp->subDTypep()};
+                    newp->dtypep(newp);
+                    if (!np) {
+                        // for Var itself
+                        nodep->dtypep(newp);
+                    } else {
+                        // for subDType
+                        np->virtRefDTypep(newp);
+                    }
+                    v3Global.rootp()->typeTablep()->addTypesp(newp);
+                    np = newp;
```

Comment:
Probably better do this inside `visit(AstUnsizedArrayDType* nodep)` similar to what is done in `visit(AstBracketArrayDType* nodep)`. But need somehow distinguish if visiting sub (sub) dtype of a `VAR` but not sub (sub) dtype of a dtype (e.g. for `STRUCTDTYPE`). Any suggestions?

Reply:
I see that `UnsizedArrayDType` stays as is for example in `typedef`. And because it is kept for dpi I think it is not OK to replace it unconditionally.

Reply:
Maybe add a bool in UnsizedArrayDType to indicate it is 'permanent'?

Reply:
But set it based on what?
I will try to get better understanding of this. At the moment I see that UnsizedArrayDType used only for `hasOpenArray...` checks for DPI. And I do not see any difference in checking for DynArrayDType there...

Reply:
I tried a few things and the results so far as below:
Placing transition to `visit(AstUnsignedArrayDType*...` without any additional checks except for DPI works OK and all tests pass, but it breaks code like
```
typedef integer da_t [];
import "DPI-C" function void some_func(input da_t i);
```
as expected, because typedef is typed as DynArrayDType.
__BUT__, actually, now, when typedef typed as UnsizedArrayDType this code is failing:
```
typedef integer da_t [];
da_t a;
...
a = da_t'{1, 2, 3, 4}; // explicit casting to typedef
```
So both ways it is somewhat not right... Still researching.

Reply:
I think that the best way to handle this is doing it in reverse order:
First create DynArrays and then convert to Unsized for DPI in the same portion of code. This way there is no need to go inside structs because it seems that open array inside struct is not allowed in DPI.

Reply:
Or change DPI in later code to accept DynArray?

Reply:
I have tried that, with no success so far. What bothers me is what if the actual DynArray is passed as _pin_ to the DPI open array _port_. But i will test this.
Also the fact that it is called _unsized array_ in the standard is making me lean towards transformation.
Marking this again as draft for now.